pod2daemon/csidriver/driver/node.go

```diff
@@ -0,0 +1,313 @@
+	return &csi.NodeUnpublishVolumeResponse{}, nil
+}
+
+func (ns *nodeService) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+	// Not needed since we are not actually creating a real volume.
+	return &csi.NodeGetVolumeStatsResponse{}, nil
+}
+
+func (ns *nodeService) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
+	// Not needed since we are not actually creating a real volume.
+	return &csi.NodeExpandVolumeResponse{}, nil
+}
+
+func (ns *nodeService) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetCapabilitiesRequest) (*csi.NodeGetCapabilitiesResponse, error) {
+	// Nothing is returned since no extra capabilities are required in order to publish/unpublish volumes.
+	return &csi.NodeGetCapabilitiesResponse{}, nil
+}
+
+func (ns *nodeService) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest) (*csi.NodeGetInfoResponse, error) {
+	// We return an empty response since this should not be needed since our "volumes" will be inline ephemeral volumes
+	// and the controller will not publish/unpublish volumes, only the node service.
+	return &csi.NodeGetInfoResponse{
+		NodeId: ns.config.NodeID,
+	}, nil
+}
+
+func extractPodInfo(req *csi.NodePublishVolumeRequest) (*creds.Credentials, error) {
+	// Extract the relevant pod info from the VolumeContext. This requires podInfoOnMount to be set for the CSI Driver object.
+	podName, podNameExists := req.VolumeContext["csi.storage.k8s.io/pod.name"]
+	podNamespace, podNamespaceExists := req.VolumeContext["csi.storage.k8s.io/pod.namespace"]
+	podUID, podUIDExists := req.VolumeContext["csi.storage.k8s.io/pod.uid"]
+	svcAcct, svcAcctExists := req.VolumeContext["csi.storage.k8s.io/serviceAccount.name"]
+
+	if !(podNameExists && podNamespaceExists && podUIDExists && svcAcctExists) {
+		return nil, fmt.Errorf("Missing the required pod info: pod.name: %s, pod.namespace: %s, pod.uid: %s, serviceAccount.name: %s", podName, podNamespace, podUID, svcAcct)
+	}
+
+	return &creds.Credentials{
+		Workload:       podName,
+		UID:            podUID,
+		Namespace:      podNamespace,
+		ServiceAccount: svcAcct,
+	}, nil
+}
+
+func (ns *nodeService) mount(destinationDir, volumeId string) error {
+	// bind destinationDir/nodeagent to /var/run/nodeagent/mount/volumeId
+	newDir := ns.config.NodeAgentWorkloadHomeDir + "/" + volumeId
+	err := os.MkdirAll(newDir, 0777)
+	if err != nil {
+		log.Errorf("Mount error: failed to create directory %s: %v", newDir, err)
+		return err
+	}
+
+	// Create the mount volume directory in /var/lib/kubelet/pods/<UID>/volumes/kubernetes.io~csi
+	err = os.MkdirAll(destinationDir, 0777)
+	if err != nil {
+		log.Errorf("Mount error: failed to create directory %s: %v", destinationDir, err)
+		return err
+	}
+
+	// Not really needed but attempt to workaround:
```

Comment:
What is this working around? The link just takes me to a file, not clear what I'm looking for there

Reply:
I'm not particularly sure, I ended up copying that from the flexvolume implementation. My understanding of it was that we are sharing the directories that k8s is mounting volumes to normally so there were probably some permission issues that the tmpfs mount overcame since the flexvolume was running on the host instead of within a pod. Now that CSI runs inside a pod, it is definitely likely we don't need this. I'll add a TODO to investigate it later.